src/writer.rs

```diff
@@ -637,36 +664,43 @@ impl<D: Distance> Writer<D> {
         n_trees: Option<usize>,
         item_indices: &RoaringBitmap,
         frozen_reader: &FrozzenReader<D>,
-    ) -> Result<(Vec<ItemId>, Vec<TmpNodesReader>)> {
+        concurrent_node_ids: ConcurrentNodeIds,
+    ) -> Result<(Vec<ItemId>, Vec<TmpNodesReader>, ConcurrentNodeIds)> {
         let n_items = item_indices.len();
-        let concurrent_node_ids = frozen_reader.concurrent_node_ids;
+        let base_concurrent_node_id = concurrent_node_ids.clone();
 
+        let ((root_nodes, tmp_nodes_readers), concurrents_node_ids) =
         repeatn(rng.next_u64(), n_trees.unwrap_or(usize::MAX))
             .enumerate()
+            .zip(repeatn(base_concurrent_node_id, usize::MAX))
             // Stop generating trees once the number of tree nodes are generated
             // but continue to generate trees if the number of trees is specified
-            .take_any_while(|_| match n_trees {
+            .take_any_while(|(_, concurrent_node_ids)| match n_trees {
                 Some(_) => true,
                 None => concurrent_node_ids.used() < n_items,
             })
-            .map(|(i, seed)| {
+            .map(|((i, seed), mut concurrent_node_ids)| {
                 log::debug!("started generating tree {i:X}...");
                 let mut rng = R::seed_from_u64(seed.wrapping_add(i as u64));
                 let mut tmp_nodes = match self.tmpdir.as_ref() {
                     Some(path) => TmpNodes::new_in(path)?,
                     None => TmpNodes::new()?,
                 };
                 let root_id =
-                    self.make_tree_in_file(frozen_reader, &mut rng, item_indices, &mut tmp_nodes)?;
+                    self.make_tree_in_file(frozen_reader, &mut rng, item_indices, &mut concurrent_node_ids, &mut tmp_nodes)?;
                 assert!(
                     root_id.mode != NodeMode::Item,
                     "make_tree_in_file returned an item even though there was more than a single element"
                 );
                 log::debug!("finished generating tree {i:X}");
                 // make_tree will NEVER return a leaf when called as root
-                Ok((root_id.unwrap_tree(), tmp_nodes.into_bytes_reader()?))
+                Ok(((root_id.unwrap_tree(), tmp_nodes.into_bytes_reader()?), concurrent_node_ids))
             })
-            .collect()
+            .collect::<Result<((Vec<ItemId>, Vec<TmpNodesReader>), Vec<ConcurrentNodeIds>)>>()?;
+
+        let concurrent_node_ids =
+            ConcurrentNodeIds::merge(concurrents_node_ids).unwrap_or(concurrent_node_ids);
```

Comment:
Can this unwrap happen?

Reply:
Yes, we don't know if we need to build more trees than what we have currently; thus, we may call `build_tree` for it to run 0 times